pkg/codegen/internal/test/helpers.go

```diff
@@ -142,28 +263,20 @@ func RewriteFilesWhenPulumiAccept(t *testing.T, dir, lang string, actual map[str
 	baseline := filepath.Join(dir, lang)
 
 	// Remove the baseline directory's current contents.
-	entries, err := os.ReadDir(baseline)
+	_, err := os.ReadDir(baseline)
 	switch {
 	case err == nil:
-		for _, e := range entries {
-			err = os.RemoveAll(filepath.Join(baseline, e.Name()))
-			require.NoError(t, err)
-		}
+		ignore, err := loadIgnoreMap(baseline)
+		require.NoError(t, err)
+		err = removeFilesFromDirUnlessIgnored(baseline, baseline, ignore)
```

Comment:
I don't understand why we would want to leave some files behind. This makes the test result dependent on local files, perhaps generated from a previous test. Shouldn't these tests be entirely driven from schema.

Reply:
Thank you for this comment it's worth thinking through and considering alternatives.

1. I find it awfully convenient to be able to add hand-authored files into the code-generated test tree, specifically test files defining unit (mock) tests over the generated code. Being in the same tree makes it easy to open up that tree in an editor and for it to feel like a normal project when editing those tests. So I would like to keep this capability.

2. The point you bring up is very important also - the rewrite procedures should completely remove files generated from a previous test, otherwise we cannot trust that we are actually checking the code-gen. Before the change, this was done by removing the folder whole-sale, which was trivially correct but conflicting with *1*. After the change, this is relying on `.sdkcodegenignore` patterns to teach the system which sub-trees to leave, assuming (but not knowing for sure) that no code-generated files would match the `.sdkcodegenignore` patterns. This also has the side-effect of leaving build files like node_modules alone when doing codegen rewrites, which is hopefully benign (can be removed by `git clean -fxd`).

How else could we accomplish 1+2?

a. Identify code-generated files by headers "// this file was generated", though this is problematic with json files

b. Make the code-generator emit a list of files "generated.txt" when generating, and clean only those

c. Slight variation on (1) where we copy hand-authored files from a "mixin" folder, such as:

```
rm -rf $codedir
mkdir $codedir
generate-code --to $codedir
cp -r $codedir-mixin/ $codedir/ # copy test files into the tree
cd $codedir && go test # etc
``` 

I personally think we might be ok as-is for now, but option (b) is attractive as it is just as ergonomic and has files in one place while being fully precise. Thoughts?

Reply:
> This also has the side-effect of leaving build files like node_modules alone when doing codegen rewrites, which is hopefully benign (can be removed by git clean -fxd).

I find this pretty worrisome from the point of view of type-checking the generated code. It feels like it's difficult to guarantee that leaving these files behind won't affect type checking. For example, if we ignore `go.mod` and it doesn't get overwritten before the next round of type checking, the results of type checking might not really be representative.

I should note that this shouldn't bite us in terms of overall validation--CI will always start from a clean state--but I do find it incredibly irritating when tests pass locally and fail in CI due to the sorts of environmental issues that are possible when we leave these things behind.

Reply:
> I should note that this shouldn't bite us in terms of overall validation--CI will always start from a clean state--but I do find it incredibly irritating when tests pass locally and fail in CI due to the sorts of environmental issues that are possible when we leave these things behind.

Yes

I'm not sure _a_ and _b_ are viable. They works for files we generate directly, but fail for files generated by other processes (like `yarn.lock`). Solving this problem was why I had the `$TEST/typecheck` folder. I would advocate for _c_. It's less ergonomic, but easier to implement and obviously correct.

Reply:
So about that.. 

`go clean -fxd` removes everything that's not checked into git, whether it's .gitignore'd or not. And even with current changes, we assume that checked in changes are intentional (and will match CI). That command is muscle memory for me and I use it a lot to make sure I'm 'testing from clean'. I'd not want to make it part of the encoded test flow because sometimes you don't want that.. for example you have dirty edits to a test file in progress or you just don't want to blow the build cache on iterative changes.

Would it help if I checked in a developer-oriented makefile in this folder with some of these commands, like `clean::` could do `git clean -fxd`; and we could have a helper command for the PULUMI_ACCEPT=1 run to make it more discoverable?

> For example, if we ignore go.mod and it doesn't get overwritten before the next round of type checking, the results of type checking might not really be representative.

`go.mod` is a special snowflake. In current version it is always overwritten by the `typeCheckGeneratedPackage` pass. Because I did not find a way to make codegen itself generate it as one of the official outputs (search for go.mod yields nothing around gen.go). So this particular example is not possible.

Reply:
Let me come back to this on Monday and make a call. If I cannot convince you, the right thing is to cede ergonomics in favor of a simpler and more robust foolproof solution.